treeTrimming.js

```diff
@@ -104,13 +108,23 @@ function trimDeadEndNodes(node){
 function trimRedundantNodes(inputTree, attribute){
 	/*call the other two tree trimming functions first, because they might create
 	redundant nodes. trimSilentTerminals() might create dead-end terminals,
-	so call that inside of trim deadEndTerminals(). trimSilentTerminals()
+	so also call trimDeadEndNodes(). trimSilentTerminals()
 	creates a copy of the tree*/
 	if(attribute=="silent"){
-		var tree = trimDeadEndNodes(trimSilentTerminals(inputTree));
+		var tree = trimSilentTerminals(inputTree);
 	}else if(attribute=="func"){
-		var tree = trimDeadEndNodes(trimFunctionalTerminals(inputTree));
+		var tree = trimFunctionalTerminals(inputTree);
+	}else{
+		tree = inputTree;
 	}
+	var i = 0;
+
+	//Must call multiple times to make sure all dead-ends are removed. If run only once, a non-x0 terminal will be removed but leave another dead end in its parent.
+	while(i < 5){
```

Comment:
Why 5 times? This number seems arbitrary. Shouldn't the max be linked to the depth of the tree?

Reply:
Alternative 1: Instead of using a constant 5, use the depth of the tree to determine the number of times to run this. You would have to write a function to determine the depth of the tree, however.

Alternative 2: Instead of using a constant 5, use a conditional that checks whether the number of leaves has changed. You would need to add an option to getLeaves that makes it category-sensitive, so that it only returns things with category x0 (for example).

Reply:
Resolved, now runs trimDeadEndNodes() if the leaves of the tree are not all category "x0."